WalletWasabi.Fluent/ViewModels/CoinSelection/SelectCoinsDialogViewModel.cs

```diff
@@ -0,0 +1,161 @@
+			: transactionInfo.MinimumRequiredAmount;
+
+		_transactionLabels = transactionInfo.UserLabels;
+
+		SetupCancel(enableCancel: false, enableCancelOnEscape: true, enableCancelOnPressed: false);
+		EnableBack = true;
+		NextCommand = ReactiveCommand.Create(() => new List<SelectableCoin>());
+	}
+
+	public Money TargetAmount { get; }
+
+	private new ReactiveCommand<Unit, List<SelectableCoin>> NextCommand { get; set; }
+
+	[SuppressMessage("Reliability", "CA2000:Dispose objects before losing scope", Justification = "Objects use DisposeWith")]
+	protected override void OnNavigatedTo(bool isInHistory, CompositeDisposable disposables)
+	{
+		var coinChanges = _balanceChanged
+			.StartWith()
+			.ObserveOn(RxApp.MainThreadScheduler)
+			.SelectMany(_ => GetCoins())
+			.ToObservableChangeSet(x => x.OutPoint)
+			.TransformWithInlineUpdate(smartCoin => new SelectableCoin(smartCoin), (selectableCoin, smartCoin) => selectableCoin.Coin = smartCoin)
+			.Replay();
+
+		coinChanges.Connect().DisposeWith(disposables);
+
+		coinChanges
+			.Bind(out var coinCollection)
+			.Subscribe()
+			.DisposeWith(disposables);
+
+		var allCoins = coinChanges
+			.AutoRefresh(x => x.IsSelected)
+			.ToCollection()
+			.Throttle(TimeSpan.FromMilliseconds(100), RxApp.MainThreadScheduler);
+
+		var selectedCoins = allCoins
+			.Select(items => items.Where(t => t.IsSelected).ToList());
+
+		EnoughSelected = selectedCoins.Select(coins => coins.Sum(x => x.Amount) >= TargetAmount);
+
+		IsSelectionBadlyChosen = allCoins.Select(coins => IsSelectionBadForPrivacy(coins, _transactionLabels)).ReplayLastActive();
+
+		SelectedAmount = selectedCoins.Select(Sum);
+
+		RemainingAmount = SelectedAmount.Select(money => Money.Max(TargetAmount - money, Money.Zero));
+
+		SelectedCount = selectedCoins.Select(models => models.Count);
+
+		NextCommand = ReactiveCommand.CreateFromObservable(() => selectedCoins, EnoughSelected);
+		NextCommand.Subscribe(models => Close(DialogResultKind.Normal, _walletViewModel.Wallet.Coins.Where(x => models.Any(coin => coin.OutPoint == x.OutPoint))));
+
+		SelectPredefinedCoinsCommand = ReactiveCommand.Create(() => coinCollection.ToList().ForEach(x => x.IsSelected = _usedCoins.Any(coin => x.Coin.OutPoint == coin.OutPoint)));
+
+		SelectAllCoinsCommand = ReactiveCommand.Create(() => coinCollection.ToList().ForEach(x => x.IsSelected = true));
+
+		ClearCoinSelectionCommand = ReactiveCommand.Create(() => coinCollection.ToList().ForEach(x => x.IsSelected = false));
+
+		SelectAllPrivateCoinsCommand = ReactiveCommand.Create(() => coinCollection.ToList().ForEach(coinViewModel => coinViewModel.IsSelected = coinViewModel.PrivacyLevel == PrivacyLevel.Private));
+
+		var commands = new[]
+		{
+			new CommandViewModel("All", SelectAllCoinsCommand),
+			new CommandViewModel("None", ClearCoinSelectionCommand),
+			new CommandViewModel("Private coins", SelectAllPrivateCoinsCommand),
+			new CommandViewModel("Smart", SelectPredefinedCoinsCommand)
+		};
```

Comment:
While the use of commands list is clever, arguably it puts dynamic behavior in a place where it shouldn't.

We don't have a "list of buttons" in the UI. We have 4 specific static buttons, which won't change regardless of any conditions, it's not like you add or remove buttons in any cases.

I would go by the simpler XAML-based approach of `<Button Command={Binding SelectAllCoinsCommand}/>` and so on

Reply:
I see your point, but this scenario is different because the commands are responsibility of class `SelectCoinsDialogViewModel`, but the commands are used in the selection header. We need to carry those commands somehow to the `ColumnFactory`. We can't put the logic outside *information expert*. I'm open to concrete solutions you might find. Maybe I'm missing something :)